[FLINK-29478][flink-connector/hive] Flink connector hive upgrade to 3.1.3

Samrat002 opened the pull request:
## What is the purpose of the change
Associated JIRA :- 
  [FLINK-29478](https://issues.apache.org/jira/browse/FLINK-29478)

Hive new version 3.1.3 is release 08 April 2022 
this change updated the `flink-connector-hive-3.1.2` to `flink-connector-hive-3.1.3`


## Brief change log
Hive new version 3.1.3 is release 08 April 2022 
this change updated the `flink-connector-hive-3.1.2` to `flink-connector-hive-3.1.3`




## Verifying this change



This change is already covered by existing e2e tests .



This change added tests and can be verified as follows:
  - Update the flink-hive-connector support 
  - Updated the e2e test to check the newly updated hive version 3.1.3
  - Tried testing for connector in EMR and validated manually. 
  - Manually verified the change by running a 5 node cluster (m5.4x instances in aws EMR) and tried basic sql commands to verify the following 
     1. create database in hive (3.1.3 ) , 
     2. create unmanaged table in hive and update table, 
     3. add data in table 
     4. truncate table 
     5. drop table

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  yes 
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: don't know
  - The runtime per-record code paths (performance sensitive): don't know
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?:  no, Up

MartijnVisser reviewed (approved):
This looks good to me but I would like to get @luoyuxia opinion too :)

MartijnVisser commented on flink-connectors/flink-connector-hive/pom.xml:
```diff
@@ -1,3 +1,4 @@
+
```
Can you remove this blank line?

Samrat002 replied:
removed the blank line.

MartijnVisser reviewed (commented):
Note: I think we should have already updated the dependencies when https://issues.apache.org/jira/browse/FLINK-23902 was completed

dannycranmer commented:
Given this change is almost a no-op, are we sure we cannot simply rename `flink-connector-hive-3.1.2` to `flink-connector-hive-3.1` and bump the patch version? I know I already asked on the Jira, but will the 3.1.3 version of the connector not work with hive 3.1.2? Seems like a lot of work for a patch version bump, especially considering they appear to be API compatbile.

MartijnVisser commented:
> but will the 3.1.3 version of the connector not work with hive 3.1.2? Seems like a lot of work for a patch version bump, especially considering they appear to be API compatbile.

I believe it's more that for each Hive version there could be a shim that needs to be introduced in Flink to make it work. It's too tightly coupled with Flink at the moment (which is why https://issues.apache.org/jira/browse/FLINK-26603 is so important)

luoyuxia reviewed (commented):
@Samrat002 Thanks for contribution. I just notice there still exists `'hive-version' = '3.1.2'` in docs. Could you please also change them to `'hive-version' = '3.1.3'` since we will offically provide flink-sql-connector-hive-3.1.3?
Also it'll be better also change  the line `hiveCatalog = HiveTestUtils.createHiveCatalog(CatalogTest.TEST_CATALOG_NAME, "3.1.2");` in `HiveTemporalJoinITCase#setup` to make the codebase consistent.

Samrat002 commented:
> hiveCatalog = HiveTestUtils.createHiveCatalog(CatalogTest.TEST_CATALOG_NAME, "3.1.2");

Thank you @luoyuxia for reviewing the changes ! 
i have made changes as per your suggestions! 
Please re-review again in your free time 😁

Samrat002 commented:
@flinkbot run azure

MartijnVisser commented:
@flinkbot run azure

Samrat002 commented:
looking into the test failure ! i will try to debug and get it fixed

luoyuxia commented:
Seems the failure is caused by FLINK-28766

MartijnVisser commented:
@flinkbot run azure

dannycranmer commented:
@Samrat002 can you please rebase and squash your commits. Please also change the commit message as per the [contribution guide](https://flink.apache.org/contributing/contribute-documentation.html#submit-your-contribution). 

Your PR description still contains the boilerplate, you can delete these bits, example:
> (For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)

Also please update the "Does this pull request potentially affect one of the following parts" section of the description

Samrat002 commented:
@flinkbot run azure

Samrat002 commented:
- made changes in commit message as per contributing guide 
- squashed all commits to one commit  
- removed boiler plate example in description of pr 
- updated "Does this pull request potentially affect one of the following parts" section of the description , 

    please review , i  have tried to update the description of PR with best of my understanding but i may have missed something or may need more appropriate description . 
    
 

@dannycranmer @MartijnVisser 

thank you @luoyuxia for calling out on [FLINK-28766](https://issues.apache.org/jira/browse/FLINK-28766) , i would have almost created a new issue 😅

dannycranmer commented:
I am just verifying the NOTCE file and found a couple of issues. We are bundling the following in our jar, however they are not explicitly called out:

> [INFO] Including org.apache.avro:avro:jar:1.8.2 in the shaded jar.
> [INFO] Including org.apache.avro:avro-mapred:jar:hadoop2:1.8.2 in the shaded jar.

They are mentioned as a transitive bundle:

> The bundled Apache Hive org.apache.hive:hive-exec dependency bundles the following dependencies under

I am following up to verify the hive-exec dependencies

dannycranmer commented:
I rebuilt `hive-exec` from source to get the shade output, I see some additional issues with the sql NOTICE file:

These are included in `hive-exec` 3.1.3
```
[INFO] Including org.apache.orc:orc-core:jar:1.5.8 in the shaded jar.
[INFO] Including org.apache.orc:orc-shims:jar:1.5.8 in the shaded jar.
[INFO] Including org.apache.commons:commons-lang3:jar:3.9 in the shaded jar.
[INFO] Including org.apache.orc:orc-tools:jar:1.5.8 in the shaded jar.
```

Out NOTICE file contains:
```
- org.apache.orc:orc-core:1.5.6
- org.apache.orc:orc-shims:1.5.6
- org.apache.orc:orc-tools:1.5.6
- org.apache.commons:commons-lang3:3.3.2
```

We list the following in our NOTICE file and I cannot see them bundled in `hive-exec`:
```
- org.apache.parquet:parquet-jackson:1.10.0
- com.esotericsoftware:reflectasm:1.10.1
```

I have not checked if these are new or existing issues. Regardless we need to fix them, please.

Samrat002 commented:
I will add the change to fix it

Samrat002 commented:
@flinkbot run azure

Samrat002 reviewed (commented):
I have tried to update the `NOTICE` based on dependency tree for the `flink-sql-connector-hive-3.1.3`

```
--- maven-dependency-plugin:3.2.0:tree (default-cli) @ flink-sql-connector-hive-3.1.3_2.12 ---
 org.apache.flink:flink-sql-connector-hive-3.1.3_2.12:jar:1.17-SNAPSHOT
 +- org.apache.flink:flink-connector-hive_2.12:jar:1.17-SNAPSHOT:compile
 |  \- org.apache.flink:flink-connector-files:jar:1.17-SNAPSHOT:compile
 |     \- org.apache.flink:flink-file-sink-common:jar:1.17-SNAPSHOT:compile
 +- org.apache.hive:hive-exec:jar:3.1.3:compile
 |  +- org.apache.hive:hive-vector-code-gen:jar:3.1.3:compile
 |  |  \- commons-lang:commons-lang:jar:2.6:compile
 |  +- org.apache.hive:hive-llap-tez:jar:3.1.3:compile
 |  |  +- org.apache.hive:hive-common:jar:3.1.3:compile
 |  |  |  +- org.apache.hive:hive-classification:jar:3.1.3:compile
 |  |  |  +- org.apache.hive:hive-storage-api:jar:2.7.0:compile
 |  |  |  +- org.apache.orc:orc-core:jar:1.5.8:compile
 |  |  |  |  +- org.apache.orc:orc-shims:jar:1.5.8:compile
 |  |  |  |  \- io.airlift:aircompressor:jar:0.10:compile
 |  |  |  +- jline:jline:jar:2.12:compile
 |  |  |  +- javax.servlet:javax.servlet-api:jar:3.1.0:compile
 |  |  |  +- org.eclipse.jetty:jetty-http:jar:9.3.20.v20170531:compile
 |  |  |  |  \- org.eclipse.jetty:jetty-util:jar:9.3.20.v20170531:compile
 |  |  |  +- org.eclipse.jetty:jetty-rewrite:jar:9.3.20.v20170531:compile
 |  |  |  |  \- org.eclipse.jetty:jetty-client:jar:9.3.20.v20170531:compile
 |  |  |  +- org.eclipse.jetty:jetty-server:jar:9.3.20.v20170531:compile
 |  |  |  |  \- org.eclipse.jetty:jetty-io:jar:9.3.20.v20170531:compile
 |  |  |  +- org.eclipse.jetty:jetty-servlet:jar:9.3.20.v20170531:compile
 |  |  |  |  \- org.eclipse.jetty:jetty-security:jar:9.3.20.v20170531:compile
 |  |  |  +- org.eclipse.jetty:jetty-webapp:jar:9.3.20.v20170531:compile
 |  |  |  |  \- org.eclipse.jetty:jetty-xml:jar:9.3.20.v20170531:compile
 |  |  |  +- org.apache.logging.log4j:log4j-web:jar:2.17.1:compile
 |  |  |  +- net.sf.jpam:jpam:jar:1.1:compile
 |  |  |  +- com.tdunning:json:jar:1.8:compile
 |  |  |  +- io.dropwizard.metrics:metrics-core:jar:3.1.0:compile
 |  |  |  +- io.dropwizard.metrics:metrics-jvm:jar:3.1.0:compile
 |  |  |  +- io.dropwizard.metrics:metrics-json:jar:3.1.0:compile
 |  |  |  +- com.github.joshelser:dropwizard-metrics-hadoop-metrics2-reporter:jar:0.1.2:compile
 |  |  |  \- javolution:javolution:jar:5.5.1:compile
 |  |  +- org.apache.hive:hive-llap-client:jar:3.1.3:compile
 |  |  |  \- org.apache.hive:hive-llap-common:jar:3.1.3:compile
 |  |  |     \- org.apache.hive:hive-serde:jar:3.1.3:compile
 |  |  |        +- org.apache.hive:hive-service-rpc:jar:3.1.3:compile
 |  |  |        +- org.apache.arrow:arrow-vector:jar:0.8.0:compile
 |  |  |        |  +- org.apache.arrow:arrow-format:jar:0.8.0:compile
 |  |  |        |  +- org.apache.arrow:arrow-memory:jar:0.8.0:compile
 |  |  |        |  +- io.netty:netty-buffer:jar:4.1.70.Final:compile
 |  |  |        |  \- io.netty:netty-common:jar:4.1.70.Final:compile
 |  |  |        +- com.carrotsearch:hppc:jar:0.7.2:compile
 |  |  |        +- com.vlkan:flatbuffers:jar:1.2.0-3f79e055:compile
 |  |  |        +- net.sf.opencsv:opencsv:jar:2.3:compile
 |  |  |        \- org.apache.parquet:parquet-hadoop-bundle:jar:1.10.0:compile
 |  |  \- org.apache.commons:commons-lang3:jar:3.3.2:compile
 |  +- org.apache.hive:hive-shims:jar:3.1.3:compile
 |  |  +- org.apache.hive.shims:hive-shims-common:jar:3.1.3:compile
 |  |  +- org.apache.hive.shims:hive-shims-0.23:jar:3.1.3:runtime
 |  |  |  \- org.apache.hadoop:hadoop-yarn-server-resourcemanager:jar:3.1.0:runtime
 |  |  |     +- org.eclipse.jetty:jetty-util-ajax:jar:9.3.19.v20170502:runtime
 |  |  |     +- org.apache.hadoop:hadoop-yarn-server-common:jar:3.1.0:runtime
 |  |  |     |  +- org.apache.geronimo.specs:geronimo-jcache_1.0_spec:jar:1.0-alpha-1:runtime
 |  |  |     |  +- org.ehcache:ehcache:jar:3.3.1:runtime
 |  |  |     |  +- com.zaxxer:HikariCP-java7:jar:2.4.12:runtime
 |  |  |     |  \- com.microsoft.sqlserver:mssql-jdbc:jar:6.2.1.jre7:runtime
 |  |  |     +- org.apache.hadoop:hadoop-yarn-server-applicationhistoryservice:jar:3.1.0:runtime
 |  |  |     |  \- de.ruedigermoeller:fst:jar:2.50:runtime
 |  |  |     |     \- com.cedarsoftware:java-util:jar:1.9.0:runtime
 |  |  |     |        \- com.cedarsoftware:json-io:jar:2.5.1:runtime
 |  |  |     +- org.apache.hadoop:hadoop-yarn-server-web-proxy:jar:3.1.0:runtime
 |  |  |     \- org.fusesource.leveldbjni:leveldbjni-all:jar:1.8:runtime
 |  |  \- org.apache.hive.shims:hive-shims-scheduler:jar:3.1.3:runtime
 |  +- commons-codec:commons-codec:jar:1.15:compile
 |  +- commons-io:commons-io:jar:2.11.0:compile
 |  +- org.antlr:ST4:jar:4.0.4:compile
 |  +- org.apache.ant:ant:jar:1.9.1:compile
 |  |  \- org.apache.ant:ant-launcher:jar:1.9.1:compile
 |  +- org.apache.commons:commons-compress:jar:1.21:compile
 |  +- org.apache.thrift:libfb303:jar:0.9.3:compile
 |  |  \- org.apache.thrift:libthrift:jar:0.9.3:compile
 |  |     +- org.apache.httpcomponents:httpclient:jar:4.5.13:compile
 |  |     \- org.apache.httpcomponents:httpcore:jar:4.4.14:compile
 |  +- org.apache.hadoop:hadoop-yarn-registry:jar:3.1.0:compile
 |  |  +- org.apache.hadoop:hadoop-auth:jar:3.1.0:compile
 |  |  |  +- com.nimbusds:nimbus-jose-jwt:jar:4.41.1:compile
 |  |  |  |  \- com.github.stephenc.jcip:jcip-annotations:jar:1.0-1:compile
 |  |  |  +- net.minidev:json-smart:jar:2.3:compile
 |  |  |  |  \- net.minidev:accessors-smart:jar:1.2:compile
 |  |  |  |     \- org.ow2.asm:asm:jar:5.0.4:compile
 |  |  |  \- org.apache.kerby:kerb-simplekdc:jar:1.0.1:compile
 |  |  |     +- org.apache.kerby:kerb-client:jar:1.0.1:compile
 |  |  |     |  +- org.apache.kerby:kerby-config:jar:1.0.1:compile
 |  |  |     |  +- org.apache.kerby:kerb-core:jar:1.0.1:compile
 |  |  |     |  |  \- org.apache.kerby:kerby-pkix:jar:1.0.1:compile
 |  |  |     |  |     +- org.apache.kerby:kerby-asn1:jar:1.0.1:compile
 |  |  |     |  |     \- org.apache.kerby:kerby-util:jar:1.0.1:compile
 |  |  |     |  +- org.apache.kerby:kerb-common:jar:1.0.1:compile
 |  |  |     |  |  \- org.apache.kerby:kerb-crypto:jar:1.0.1:compile
 |  |  |     |  +- org.apache.kerby:kerb-util:jar:1.0.1:compile
 |  |  |     |  \- org.apache.kerby:token-provider:jar:1.0.1:compile
 |  |  |     \- org.apache.kerby:kerb-admin:jar:1.0.1:compile
 |  |  |        +- org.apache.kerby:kerb-server:jar:1.0.1:compile
 |  |  |        |  \- org.apache.kerby:kerb-identity:jar:1.0.1:compile
 |  |  |        \- org.apache.kerby:kerby-xdr:jar:1.0.1:compile
 |  |  +- org.apache.hadoop:hadoop-yarn-api:jar:3.1.0:compile
 |  |  +- org.apache.hadoop:hadoop-common:jar:2.8.5:compile
 |  |  |  +- org.apache.commons:commons-math3:jar:3.6.1:compile
 |  |  |  +- xmlenc:xmlenc:jar:0.52:compile
 |  |  |  +- commons-collections:commons-collections:jar:3.2.2:compile
 |  |  |  +- org.mortbay.jetty:jetty-sslengine:jar:6.1.26:compile
 |  |  |  +- javax.servlet.jsp:jsp-api:jar:2.1:runtime
 |  |  |  +- net.java.dev.jets3t:jets3t:jar:0.9.0:compile
 |  |  |  |  \- com.jamesmurty.utils:java-xmlbuilder:jar:0.4:compile
 |  |  |  +- commons-configuration:commons-configuration:jar:1.7:compile
 |  |  |  |  +- commons-digester:commons-digester:jar:1.8.1:compile
 |  |  |  |  \- commons-beanutils:commons-beanutils:jar:1.8.3:compile
 |  |  |  +- com.jcraft:jsch:jar:0.1.54:compile
 |  |  |  \- org.apache.htrace:htrace-core4:jar:4.0.1-incubating:compile
 |  |  +- org.apache.curator:curator-client:jar:2.12.0:compile
 |  |  +- org.apache.curator:curator-recipes:jar:2.12.0:compile
 |  |  +- commons-cli:commons-cli:jar:1.5.0:compile
 |  |  +- commons-daemon:commons-daemon:jar:1.0.13:compile
 |  |  +- commons-net:commons-net:jar:3.6:compile
 |  |  \- dnsjava:dnsjava:jar:2.1.7:compile
 |  +- org.apache.ivy:ivy:jar:2.4.0:compile
 |  +- org.apache.zookeeper:zookeeper:jar:3.5.9:compile
 |  |  +- org.apache.zookeeper:zookeeper-jute:jar:3.5.9:compile
 |  |  \- org.apache.yetus:audience-annotations:jar:0.5.0:compile
 |  +- org.apache.curator:curator-framework:jar:2.12.0:compile
 |  +- org.apache.curator:apache-curator:pom:2.12.0:compile
 |  +- org.codehaus.groovy:groovy-all:jar:2.4.11:compile
 |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.13.4:compile
 |  +- com.fasterxml.jackson.core:jackson-core:jar:2.13.4:compile
 |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.13.4:compile
 |  +- org.datanucleus:datanucleus-core:jar:4.1.17:compile
 |  +- org.apache.calcite:calcite-core:jar:1.16.0:compile
 |  |  +- org.apache.calcite:calcite-linq4j:jar:1.16.0:compile
 |  |  +- commons-dbcp:commons-dbcp:jar:1.4:compile
 |  |  |  \- commons-pool:commons-pool:jar:1.5.4:compile
 |  |  +- com.esri.geometry:esri-geometry-api:jar:2.0.0:compile
 |  |  +- com.yahoo.datasketches:sketches-core:jar:0.9.0:compile
 |  |  |  \- com.yahoo.datasketches:memory:jar:0.9.0:compile
 |  |  +- org.codehaus.janino:janino:jar:2.7.6:compile
 |  |  \- org.codehaus.janino:commons-compiler:jar:2.7.6:compile
 |  +- org.apache.calcite:calcite-druid:jar:1.16.0:compile
 |  |  \- joda-time:joda-time:jar:2.5:compile
 |  +- org.apache.calcite.avatica:avatica:jar:1.11.0:compile
 |  +- com.google.code.gson:gson:jar:2.2.4:compile
 |  +- stax:stax-api:jar:1.0.1:compile
 |  \- org.apache.hive:hive-upgrade-acid:jar:3.1.3:compile
 +- org.antlr:antlr-runtime:jar:3.5.2:compile
 +- org.apache.hadoop:hadoop-mapreduce-client-core:jar:3.1.0:compile
 |  +- org.apache.hadoop:hadoop-yarn-client:jar:2.8.5:compile
 |  |  +- com.google.guava:guava:jar:11.0.2:compile
 |  |  \- commons-logging:commons-logging:jar:1.1.3:compile
 |  +- org.apache.hadoop:hadoop-yarn-common:jar:2.8.5:compile
 |  |  +- javax.xml.bind:jaxb-api:jar:2.3.1:compile
 |  |  |  \- javax.activation:javax.activation-api:jar:1.2.0:compile
 |  |  +- javax.servlet:servlet-api:jar:2.5:compile
 |  |  +- org.mortbay.jetty:jetty-util:jar:6.1.26:compile
 |  |  +- com.sun.jersey:jersey-core:jar:1.9:compile
 |  |  +- com.sun.jersey:jersey-client:jar:1.9:compile
 |  |  +- org.codehaus.jackson:jackson-jaxrs:jar:1.9.13:compile
 |  |  +- org.codehaus.jackson:jackson-xc:jar:1.9.13:compile
 |  |  +- com.google.inject:guice:jar:3.0:compile
 |  |  |  +- javax.inject:javax.inject:jar:1:compile
 |  |  |  \- aopalliance:aopalliance:jar:1.0:compile
 |  |  +- com.sun.jersey:jersey-server:jar:1.9:compile
 |  |  |  \- asm:asm:jar:3.1:compile
 |  |  +- com.sun.jersey:jersey-json:jar:1.9:compile
 |  |  |  +- org.codehaus.jettison:jettison:jar:1.1:compile
 |  |  |  \- com.sun.xml.bind:jaxb-impl:jar:2.2.3-1:compile
 |  |  \- com.sun.jersey.contribs:jersey-guice:jar:1.9:compile
 |  +- org.apache.hadoop:hadoop-hdfs-client:jar:3.1.0:compile
 |  |  \- com.squareup.okhttp:okhttp:jar:2.7.5:compile
 |  |     \- com.squareup.okio:okio:jar:1.6.0:compile
 |  +- com.google.protobuf:protobuf-java:jar:2.5.0:compile
 |  +- org.apache.hadoop:hadoop-annotations:jar:3.1.0:compile
 |  +- com.google.inject.extensions:guice-servlet:jar:4.0:compile
 |  \- io.netty:netty:jar:3.10.5.Final:compile
 +- org.apache.avro:avro:jar:1.8.2:compile
 |  +- org.codehaus.jackson:jackson-core-asl:jar:1.9.13:compile
 |  +- org.codehaus.jackson:jackson-mapper-asl:jar:1.9.13:compile
 |  +- com.thoughtworks.paranamer:paranamer:jar:2.7:compile
 |  +- org.xerial.snappy:snappy-java:jar:1.1.8.3:compile
 |  \- org.tukaani:xz:jar:1.5:compile
 +- org.apache.avro:avro-mapred:jar:hadoop2:1.8.2:compile
 |  \- org.apache.avro:avro-ipc:jar:1.8.2:compile
 |     +- org.mortbay.jetty:jetty:jar:6.1.26:compile
 |     \- org.apache.velocity:velocity:jar:1.7:compile
 +- org.slf4j:slf4j-api:jar:1.7.32:provided
 +- com.google.code.findbugs:jsr305:jar:1.3.9:provided
 +- org.apache.flink:flink-test-utils-junit:jar:1.17-SNAPSHOT:test
 |  \- org.testcontainers:testcontainers:jar:1.17.2:test
 |     +- org.rnorth.duct-tape:duct-tape:jar:1.0.8:test
 |     |  \- org.jetbrains:annotations:jar:17.0.0:test
 |     +- com.github.docker-java:docker-java-api:jar:3.2.13:test
 |     \- com.github.docker-java:docker-java-transport-zerodep:jar:3.2.13:test
 |        +- com.github.docker-java:docker-java-transport:jar:3.2.13:test
 |        \- net.java.dev.jna:jna:jar:5.8.0:test
 +- org.apache.flink:flink-shaded-force-shading:jar:15.0:compile
 +- org.junit.jupiter:junit-jupiter:jar:5.9.1:test
 |  +- org.junit.jupiter:junit-jupiter-api:jar:5.9.1:test
 |  |  +- org.opentest4j:opentest4j:jar:1.2.0:test
 |  |  \- org.junit.platform:junit-platform-commons:jar:1.9.1:test
 |  +- org.junit.jupiter:junit-jupiter-params:jar:5.9.1:test
 |  \- org.junit.jupiter:junit-jupiter-engine:jar:5.9.1:test
 +- org.junit.vintage:junit-vintage-engine:jar:5.9.1:test
 |  +- org.junit.platform:junit-platform-engine:jar:1.9.1:test
 |  +- junit:junit:jar:4.13.2:test
 |  \- org.apiguardian:apiguardian-api:jar:1.1.2:test
 +- org.assertj:assertj-core:jar:3.23.1:test
 |  \- net.bytebuddy:byte-buddy:jar:1.10.14:test
 +- org.mockito:mockito-inline:jar:3.4.6:test
 +- org.mockito:mockito-core:jar:3.4.6:test
 |  +- net.bytebuddy:byte-buddy-agent:jar:1.10.14:test
 |  \- org.objenesis:objenesis:jar:2.1:test
 +- org.mockito:mockito-junit-jupiter:jar:3.4.6:test
 +- org.powermock:powermock-module-junit4:jar:2.0.9:test
 |  +- org.powermock:powermock-module-junit4-common:jar:2.0.9:test
 |  |  +- org.powermock:powermock-reflect:jar:2.0.9:test
 |  |  \- org.powermock:powermock-core:jar:2.0.9:test
 |  |     \- org.javassist:javassist:jar:3.24.0-GA:test
 |  \- org.hamcrest:hamcrest-core:jar:1.3:test
 +- org.powermock:powermock-api-mockito2:jar:2.0.9:test
 |  \- org.powermock:powermock-api-support:jar:2.0.9:test
 +- org.hamcrest:hamcrest-all:jar:1.3:test
 +- org.testcontainers:junit-jupiter:jar:1.17.2:test
 +- org.apache.logging.log4j:log4j-slf4j-impl:jar:2.17.1:test
 +- org.apache.logging.log4j:log4j-api:jar:2.17.1:test
 +- org.apache.logging.log4j:log4j-core:jar:2.17.1:test
 \- org.apache.logging.log4j:log4j-1.2-api:jar:2.17.1:test

```

Samrat002 commented on flink-connectors/flink-sql-connector-hive-3.1.3/pom.xml:
```diff
@@ -142,18 +149,19 @@ under the License.
+										<exclude>avro/shaded/**</exclude>
 										<exclude>org/apache/avro/**</exclude>
 										<exclude>META-INF/maven/org.apache.avro/**</exclude>
+										<exclude>org/apache/hadoop/hive/conf/HiveConf*</exclude>
```
Adding these changes removes all the warning in the build. 

```
[INFO] Excluding org.apache.flink:flink-shaded-force-shading:jar:15.0 from the shaded jar.
[INFO] Replacing original artifact with shaded artifact.
[INFO] Replacing /code/flink/flink-connectors/flink-sql-connector-hive-3.1.3/target/flink-sql-connector-hive-3.1.3_2.12-1.17-SNAPSHOT.jar with /code/flink/flink-connectors/flink-sql-connector-hive-3.1.3/target/flink-sql-connector-hive-3.1.3_2.12-1.17-SNAPSHOT-shaded.jar
[INFO] Dependency-reduced POM written at: /code/flink/flink-connectors/flink-sql-connector-hive-3.1.3/target/dependency-reduced-pom.xml
[INFO] 
[INFO] --- maven-surefire-plugin:3.0.0-M5:test (integration-tests) @ flink-sql-connector-hive-3.1.3_2.12 ---
[INFO] 
[INFO] --- japicmp-maven-plugin:0.16.0_m325:cmp (default) @ flink-sql-connector-hive-3.1.3_2.12 ---
[INFO] Skipping execution because parameter 'skip' was set to true.
[INFO] 
[INFO] --- maven-install-plugin:2.5.2:install (default-install) @ flink-sql-connector-hive-3.1.3_2.12 ---
[INFO] Installing /code/flink/flink-connectors/flink-sql-connector-hive-3.1.3/target/flink-sql-connector-hive-3.1.3_2.12-1.17-SNAPSHOT.jar to /.m2/repository/org/apache/flink/flink-sql-connector-hive-3.1.3_2.12/1.17-SNAPSHOT/flink-sql-connector-hive-3.1.3_2.12-1.17-SNAPSHOT.jar
[INFO] Installing /code/flink/flink-connectors/flink-sql-connector-hive-3.1.3/target/dependency-reduced-pom.xml to /.m2/repository/org/apache/flink/flink-sql-connector-hive-3.1.3_2.12/1.17-SNAPSHOT/flink-sql-connector-hive-3.1.3_2.12-1.17-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 26.645 s
[INFO] Finished at: 2022-10-14T01:40:12+05:30
[INFO] ------------------------------------------------------------------------

```

dannycranmer commented:
I verified again and found the `flink-connector-hive` looks ok, but still some issues with the `sql` connector.

We are listing the following in our NOTICE:
```
- org.apache.hadoop:hadoop-mapreduce-client-core:3.1.0
```

Yes these are not bundled:

```
[INFO] Excluding org.apache.hadoop:hadoop-mapreduce-client-core:jar:3.1.0 from the shaded jar
```

Regarding "The bundled Apache Hive org.apache.hive:hive-exec dependency" section, I am not sure this is correct. I see we have dropped guava from the list, however I can see guava is bundled:

```
unzip -l target/flink-sql-connector-hive-3.1.3_2.12-1.17-SNAPSHOT.jar | grep google
...
19380  10-14-2022 09:53   avro/shaded/com/google/common/util/concurrent/Futures.class
...
```

Samrat002 commented:
@dannycranmer 

I removed `org.apache.hadoop:hadoop-mapreduce-client-core:3.1.0` from `NOTICE`.

> Regarding "The bundled Apache Hive org.apache.hive:hive-exec dependency" section, I am not sure this is correct. I see we have dropped guava from the list, however I can see guava is bundled:

i found this in [pom.xml](https://github.com/apache/flink/blob/master/flink-connectors/flink-sql-connector-hive-3.1.2/pom.xml#L167). Not sure if this is relevent

dannycranmer commented on flink-connectors/flink-sql-connector-hive-3.1.3/pom.xml:
```diff
@@ -142,18 +149,19 @@ under the License.
 									<excludes>
-										<exclude>org/apache/hadoop/hive/conf/HiveConf.class</exclude>
-										<exclude>org/apache/hadoop/hive/metastore/HiveMetaStoreClient.class</exclude>
+										<exclude>avro/shaded/**</exclude>
```
Why did you add this?

Samrat002 replied:
Adding `<exclude>avro/shaded/**</exclude>` removes the following warning in compilation of `flink-sql-connector-hive-3.1.3`

```
[WARNING] hive-exec-3.1.3.jar, avro-1.8.2.jar define 799 overlapping classes: 
[WARNING]   - avro.shaded.com.google.common.collect.SingletonImmutableList
[WARNING]   - avro.shaded.com.google.common.collect.Iterables$15
[WARNING]   - avro.shaded.com.google.common.collect.Sets$PowerSet$1$1
[WARNING]   - avro.shaded.com.google.common.collect.RegularImmutableMap
[WARNING]   - avro.shaded.com.google.common.collect.Synchronized$SynchronizedSortedSet
[WARNING]   - avro.shaded.com.google.common.collect.Synchronized$SynchronizedObject
[WARNING]   - avro.shaded.com.google.common.cache.LocalCache$Strength
[WARNING]   - avro.shaded.com.google.common.collect.ForwardingMultiset
[WARNING]   - avro.shaded.com.google.common.collect.MapMakerInternalMap$WeakExpirableEntry
[WARNING]   - avro.shaded.com.google.common.collect.Collections2$FilteredCollection$1
[WARNING]   - 789 more...
[WARNING] maven-shade-plugin has detected that some class files are
[WARNING] present in two or more JARs. When this happens, only one
[WARNING] single version of the class is copied to the uber jar.
[WARNING] Usually this is not harmful and you can skip these warnings,
[WARNING] otherwise try to manually exclude artifacts based on
[WARNING] mvn dependency:tree -Ddetail=true and the above output.
[WARNING] See http://maven.apache.org/plugins/maven-shade-plugin/

```

dannycranmer commented:
>i found explicit message
in [pom.xml](https://github.com/apache/flink/blob/master/flink-connectors/flink-sql-connector-hive-3.1.2/pom.xml#L167). Not sure if this is relevent

Seems like it might be wrong, regardless we ARE bundling Guava into the jar and it is not mentioned in the NOTICE file. Before we start removing shaded dependencies we need to be certain these are not needed for the SQL client

Samrat002 commented:
@flinkbot run azure

dannycranmer commented on flink-connectors/flink-sql-connector-hive-3.1.3/pom.xml:
```diff
@@ -164,11 +172,6 @@ under the License.
 									<pattern>org.apache.parquet</pattern>
 									<shadedPattern>org.apache.hive.shaded.parquet</shadedPattern>
 								</relocation>
-								<!-- Guava dependencies that still end up in the shaded JAR are relocated to avoid clashes. -->
-								<relocation>
-									<pattern>com.google</pattern>
-									<shadedPattern>org.apache.flink.hive.shaded.com.google</shadedPattern>
-								</relocation>
```
This is needed, I can now see we are including 2x versions of Guava into the jar:
- The non-relocated version (com.google)
- The version coming from avro (avro.shaded)

```
unzip -l target/flink-sql-connector-hive-3.1.3_2.12-1.17-SNAPSHOT.jar | grep /Futures.class
    25448  10-18-2022 20:27   com/google/common/util/concurrent/Futures.class
    19380  10-18-2022 20:27   avro/shaded/com/google/common/util/concurrent/Futures.class
```

Samrat002 replied:
reverted the changes.



 ```
 unzip -l target/flink-sql-connector-hive-3.1.3_2.12-1.17-SNAPSHOT.jar | grep /Futures.class     
    31886  10-19-2022 14:36   org/apache/flink/hive/shaded/com/google/common/util/concurrent/Futures.class
    19380  10-19-2022 14:36   avro/shaded/com/google/common/util/concurrent/Futures.class
```

dannycranmer commented on flink-connectors/flink-sql-connector-hive-3.1.3/src/main/resources/META-INF/NOTICE:
```diff
@@ -0,0 +1,52 @@
+- org.antlr:antlr-runtime:3.5.2
+
+The bundled Apache Hive org.apache.hive:hive-exec dependency bundles the following dependencies under
+the Apache Software License 2.0 (http://www.apache.org/licenses/LICENSE-2.0.txt)
```
We still do not have Guava in the NOTICE. Shall we jump on a call tomorrow and discuss? I feel like we are not making progress. I will slack you.

Samrat002 replied:
Thank you @dannycranmer for helping me out in understanding the `NOTICE` related things specific to flink

dannycranmer commented:
@flinkbot run azure

dannycranmer commented on flink-connectors/flink-sql-connector-hive-3.1.3/src/main/resources/META-INF/NOTICE:
```diff
@@ -25,25 +26,22 @@ the Apache Software License 2.0 (http://www.apache.org/licenses/LICENSE-2.0.txt)
 - javax.jdo:jdo-api:3.0.1
 - joda-time:joda-time:2.9.9
 - net.sf.opencsv:opencsv:2.3
-- org.apache.avro:avro-mapred:hadoop2:1.8.2
-- org.apache.avro:avro:1.8.2
```
I believe we should retain these, since they are bundled into hive-exec also:

Extract from `hive-exec` build
```
[INFO] Including org.apache.avro:avro:jar:1.8.2 in the shaded jar.
[INFO] Including org.apache.avro:avro-mapred:jar:hadoop2:1.8.2 in the shaded jar.
```

dannycranmer replied:
ok, the build does not allow the dependency to be defined twice:

```
2022-10-19T10:53:16.1186081Z 		 These issue are legally problematic and MUST be fixed: 
2022-10-19T10:53:16.1186514Z 			Dependency org.apache.avro:avro:1.8.2 is declared twice.
2022-10-19T10:53:16.1187176Z 			Dependency org.apache.avro:avro-mapred:hadoop2:1.8.2 is declared twice.
```

dannycranmer reviewed (changes requested):
NOTICE file looks good to me, minus the one comment I left.

Samrat002 commented:
@flinkbot run azure

dannycranmer commented:
@flinkbot run azure

MartijnVisser commented:
CI has passed in https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=42229&view=results - Merging

MartijnVisser commented:
@Samrat002 One minor note for next time: don't forget the naming conventions of commit messages. I've changed it when I merged it for now. See https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html#commit-naming-conventions

KarlManong commented:
@Samrat002 will this patch port to 1.15, I see https://issues.apache.org/jira/browse/FLINK-28729 reported

MartijnVisser commented:
@KarlManong No, we don't do these type of changes in patch releases.
